exporter/elasticsearchexporter/elasticsearch_bulk.go

```diff
@@ -175,52 +162,160 @@ func createElasticsearchBackoffFunc(config *RetrySettings) func(int) time.Durati
 	}
 }
 
-func shouldRetryEvent(status int, retryOnStatus []int) bool {
-	for _, retryable := range retryOnStatus {
-		if status == retryable {
-			return true
+func pushDocuments(ctx context.Context, index string, document []byte, bulkIndexer *esBulkIndexerCurrent) error {
+	return bulkIndexer.Add(ctx, index, bytes.NewReader(document))
+}
+
+func newBulkIndexer(logger *zap.Logger, client *elasticsearch7.Client, config *Config) (*esBulkIndexerCurrent, error) {
+	numWorkers := config.NumWorkers
+	if numWorkers == 0 {
+		numWorkers = runtime.NumCPU()
+	}
+
+	flushInterval := config.Flush.Interval
+	if flushInterval == 0 {
+		flushInterval = 30 * time.Second
+	}
+
+	flushBytes := config.Flush.Bytes
+	if flushBytes == 0 {
+		flushBytes = 5e+6
+	}
+
+	var maxDocRetry int
+	if config.Retry.Enabled {
+		// max_requests includes initial attempt
+		// See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32344
+		maxDocRetry = config.Retry.MaxRequests - 1
+	}
+
+	pool := &bulkIndexerPool{
+		wg:    sync.WaitGroup{},
+		items: make(chan esBulkIndexerItem, config.NumWorkers),
+		stats: bulkIndexerStats{},
+	}
+	pool.wg.Add(numWorkers)
+
+	for i := 0; i < numWorkers; i++ {
+		bi, err := docappender.NewBulkIndexer(docappender.BulkIndexerConfig{
+			Client:                client,
+			MaxDocumentRetries:    maxDocRetry,
+			CompressionLevel:      0,
```

Comment:
- Used 0 to avoid an import, but I agree that gzip.NoCompression will be clearer

> Also, I see that this field is not exposed as a config, what is the reasoning behind going with no compression?

I do not want to introduce any behavioral change on compression in this PR. It can be done in a follow up PR to expose compression config.

Reply:
> Used 0 to avoid an import, especially when docappender does not use the stock gzip

We can just not set the `CompressionLevel` in that case (a lot clearer IMO)

Also, docappender uses “github.com/klauspost/compress/gzip”` import but it should already be in the indirect dependency tree.